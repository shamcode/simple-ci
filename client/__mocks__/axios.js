const METHODS = [
    'get',
    'post',
    'put',
    'delete'
];

const FIXTURES = {};
const MOCKS = {};

function mockMethodFactory( method ) {
    return jest.fn().mockImplementation(
        ( url ) => {
            if ( !FIXTURES[ method ].hasOwnProperty( url ) ) {
                throw new Error( `Missing fixture for ${method.toUpperCase()} url = "${url}"` );
            }
            const { success, data } = FIXTURES[ method ][ url ];
            return Promise[ success ? 'resolve' : 'reject' ]( {
                data
            } );
        }
    );
}

METHODS.forEach(
    method => {
        MOCKS[ method ] = mockMethodFactory( method );
        FIXTURES[ method ] = {};
    }
);

const defaultMocksData = ( () => {
    const data = {};
    data.chain = {
        id: 2,
        name: 'Test chain',
        command: 'echo 1'
    };
    data.project = {
        id: 1,
        name: 'Test',
        cwd: '/tmp/',
        chains: [ data.chain ]
    };
    return data;
} )();

export default {
    defaultMocksData: defaultMocksData,
    mocks: MOCKS,
    create: jest.fn().mockImplementation( () => {
        return {
            ...MOCKS,
            interceptors: {
                request: {
                    use: () => {}
                },
                response: {
                    use: () => {}
                }
            }
        };
    } ),
    use( method, url, data, success = true ) {
        if ( !METHODS.includes( method ) ) {
            throw new Error( `Unknown method: "${method}"`);
        }
        FIXTURES[ method ][ url ] = {
            data,
            success
        };
        return this;
    },
    useDefaultMocks() {
        this
            .use( 'get', '/projects', [ defaultMocksData.project ] )
            .use( 'get', '/projects/1', defaultMocksData.project )
            .use( 'get',  '/chains/2', defaultMocksData.chain );
        return this;
    }
};