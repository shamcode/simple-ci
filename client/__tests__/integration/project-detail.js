import pretty from 'pretty';
import axios from 'axios';
import ShamUI, { DI } from 'sham-ui';
import controller from '../../src/controllers/main';
jest.mock( 'axios' );

const flushPromises = () => new Promise( resolve => setImmediate( resolve ) );

beforeEach( () => {
    jest.resetModules();
    jest.clearAllMocks();
} );

it( 'project-detail page', async() => {
    expect.assertions( 2 );

    const projectData = {
        id: 1,
        name: 'Test',
        cwd: '/tmp/'
    };

    axios.create.mockImplementation( () => {
        return {
            get: jest.fn()
                .mockReturnValueOnce( Promise.resolve( { data: [ projectData ] } ) )
                .mockReturnValueOnce( Promise.resolve( { data: projectData } ) ),
            interceptors: {
                response: {
                    use: () => {}
                }
            }
        };
    } );

    window.location.href = 'http://simple-ci.example.com';

    DI.bind( 'widget-binder', controller );
    const UI = new ShamUI();
    UI.render.FORCE_ALL();
    await flushPromises();

    const body = document.querySelector( 'body' );
    body.querySelector( '.project-card' ).click();
    expect( pretty( body.innerHTML ) ).toMatchSnapshot();

    await flushPromises();

    expect( pretty( body.innerHTML ) ).toMatchSnapshot();
} );

