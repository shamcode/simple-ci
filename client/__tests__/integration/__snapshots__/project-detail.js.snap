// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project-detail page 1`] = `
"<header>
  <a href=\\"\\" class=\\"logo\\">Simple CI</a>
</header>
<!--Header-->
<main>
  <div class=\\"page-data-loading\\">
    <div class=\\"sk-cube-grid\\">
      <div class=\\"sk-cube sk-cube1\\"></div>
      <div class=\\"sk-cube sk-cube2\\"></div>
      <div class=\\"sk-cube sk-cube3\\"></div>
      <div class=\\"sk-cube sk-cube4\\"></div>
      <div class=\\"sk-cube sk-cube5\\"></div>
      <div class=\\"sk-cube sk-cube6\\"></div>
      <div class=\\"sk-cube sk-cube7\\"></div>
      <div class=\\"sk-cube sk-cube8\\"></div>
      <div class=\\"sk-cube sk-cube9\\"></div>
    </div>
  </div>
  <!--PageDataLoading-->
  <!--if-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;

exports[`project-detail page 2`] = `
"<header>
  <a href=\\"\\" class=\\"logo\\">Simple CI</a>
</header>
<!--Header-->
<main>
  <!--if-->
  <!--ErrorsPanel-->
  <div class=\\"project-info\\">
    <h3>Test</h3>
    <dl>
      <dt>/tmp/</dt>
      <dd>Current work directory.</dd>
      <dd>This directory will be used when running tasks.</dd>
      <dt>
        <a href=\\"projects/1/edit\\" class=\\"\\">Edit</a>
      </dt>
      <dd>Edit project name &amp; cwd</dd>
    </dl>
  </div>
  <!--Info-->
  <!--if-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;
